### PR TITLE
Refactor string collects to itertools join

### LINKED ISSUE #1014: Replace `[T]::join` where `T: AsRef<str>` by Itertools' `join`
There's a few places where we collect an iter into a `Vec` and then join them into one big string. With itertools' `join` method on iterators, that allocation can be avoided.

### PR DESCRIPTION
**Pull request by imjacobclark**
Potentially refactored several other joins that _may_ not have been in the scope of this issue, happy to weed those out if they shouldn't be covered in this PR.

This resolves #1014.

### CODE REVIEW

File: src/types.rs [hunk 1]
```diff
@@ -477,10 +479,11 @@ impl Rewrite for ast::PolyTraitRef {
...
         if !self.bound_lifetimes.is_empty() {
             let lifetime_str = try_opt!(self.bound_lifetimes
-                    .iter()
-                    .map(|lt| lt.rewrite(context, width, offset))
-                    .collect::<Option<Vec<_>>>())
-                .join(", ");
+                .iter()
+                .map(|lt| lt.rewrite(context, width, offset))
+                .intersperse(Some(", ".to_string()))
+                .collect::<Option<String>>());
```
Comment (marcusklaas) [inline src/types.rs]: Are the type ascriptions here necessary?

File: src/items.rs [hunk 2]
```diff
@@ -1056,7 +1057,6 @@ pub fn rewrite_associated_type(ident: ast::Ident,
         let bounds: &[_] = &ty_param_bounds;
         let bound_str = bounds.iter()
             .filter_map(|ty_bound| ty_bound.rewrite(context, context.config.max_width, indent))
-            .collect::<Vec<String>>()
```
Comment (marcusklaas) [inline src/items.rs]: Actually, I think this was a bug! Could you refactor this into an intersperse -> option collect -> try_opt!, like you did with the others?

File: src/items.rs [hunk 3]
```diff
@@ -1702,7 +1702,6 @@ fn rewrite_trait_bounds(context: &RewriteContext,
 
     let bound_str = bounds.iter()
         .filter_map(|ty_bound| ty_bound.rewrite(&context, width, indent))
-        .collect::<Vec<String>>()
```
Comment (marcusklaas) [inline src/items.rs]: Idem. We should abort when a rewrite fails!

File: src/types.rs (hunk 1 above)
Comment (imjacobclark) [inline src/types.rs]: Appears so, without them we get a compiler error:

```rust
[jacobclark:~/Desktop/rustfmt]$ cargo test                                                                                             (1014-refactoring-vector-string-to-join✱) 
   Compiling rustfmt v0.5.0 (file:///Users/jacobclark/Desktop/rustfmt)
src/types.rs:488:32: 488:50 error: the type of this value must be known in this context
src/types.rs:488             let extra_offset = lifetime_str.len() + 6;
                                                ^~~~~~~~~~~~~~~~~~
src/types.rs:488:32: 488:50 error: the type of this value must be known in this context
src/types.rs:488             let extra_offset = lifetime_str.len() + 6;
                                                ^~~~~~~~~~~~~~~~~~
```

Comment (marcusklaas) [inline src/types.rs]: Thanks for checking!

**Comment by marcusklaas** [comment]
Excellent first pull request! Great initiative :-)

Could you replace the existing `filter_map`s? It's ready to go afterwards.

File: src/types.rs (hunk 1 above)
Comment (imjacobclark) [inline src/types.rs]: Update to this, they're not actually needed for each one, so removing them now to clean this up a little.

**Comment by imjacobclark** [comment]
@marcusklaas Thanks! The above is nicely refactored away, should be ready to ship!

File: src/types.rs [hunk 4]
```diff
@@ -325,39 +326,40 @@ impl Rewrite for ast::WherePredicate {
                 let type_str = try_opt!(bounded_ty.rewrite(context, width, offset));
 
                 if !bound_lifetimes.is_empty() {
-                    let lifetime_str = try_opt!(bound_lifetimes.iter()
+                    let lifetime_str: String = try_opt!(bound_lifetimes.iter()
                                                                .map(|lt| {
```
Comment (marcusklaas) [inline src/types.rs]: This indentation seems off to me.. may have to look into it. Outside the scope of this PR, though.

**Comment by marcusklaas** [comment]
:+1: Thanks for the PR!

**Pull request merged**